src/cr/cube/cube.py

```diff
@@ -773,9 +773,21 @@ def _flat_values(self):
         {'?': -1} in the cube response. These are replaced by np.nan in the
         returned value.
         """
-        return tuple(
-            np.nan if type(x) is dict else x
-            for x in self._cube_dict["result"]["measures"]["mean"]["data"]
+
+        def dict_to_nan_recursive(x):
+            if type(x) is dict:
+                return np.nan
+            elif type(x) is list:
+                return [dict_to_nan_recursive(y) for y in x]
```

Comment:
@scanny - is this the right way to do this?

The problem I hit is that because of numeric arrays, this can have two different forms:

Mean of non-array: list of floats or dictionaries (indicating missing values)
Mean of array: **list of** lists of floats or dictionaries

Reply:
I'd say those are distinct enough to warrant two different subclasses, where the factory chooses between `_MeanMeasure` and perhaps something like `_NumArrayMeanMeasure`. Then each `._flat_values` can do its own thing and we avoid using `if` statements as a substitute for polymorphism.

Btw, The docstring says "Return tuple ..." and the return value here is `ndarray` (which I think is preferable).

This issue came up in my review of Ernesto's branch here: https://github.com/Crunch-io/crunch-cube/pull/251#discussion_r586703412 where I explained that I thing _all_ `._flat_values` return type should be `ndarray` to avoid the unnecessary time and memory required to construct an intermediate `tuple` object that just gets immediately converted to an `ndarray` anyway.

So maybe a conflict brewing here, but that happens and Ernesto can figure it out during rebase (although it might be worth a mention to him in Slack). In any case, I think the right solution is to create the `_NumArrayMeanMeasure` or whatever name is appropriate and then adjust the factory in `_Measures` or wherever it is to choose the right subclass when constructing the measure. Also that these return `ndarray` and preferably that _all_ of the four subtypes return `ndarray` and the `_BaseMeasure.raw_cube_array` property is modified to avoid constructing an array out of an array (which it looks like you already did).

Also the docstring for `_WeightedCountMeasure._flat_values` also says `tuple`.

Reply:
> I'd say those are distinct enough to warrant two different subclasses, where the factory chooses between _MeanMeasure and perhaps something like _NumArrayMeanMeasure. Then each ._flat_values can do its own thing and we avoid using if statements as a substitute for polymorphism.

Just saw your new set of comments, but the problem I'm hitting as I thought about this is that I don't see a good way to distinguish numeric arrays from not at the time of the factory. Did you have thoughts on that?

My first idea was to use use `_BaseMeasure._shape` but the shape of a CAT(N categories) X Numeric is the same as a Numeric Array (N subvariables).

The other areas where the code makes decisions about numeric arrays (eg `_numeric_array_dimension`) are in the `Cube` class, but I don't see a good way to pass that information along when creating the measures.